### ISSUE TITLE
DataFusion coercion logic is different from Arrow's

### ISSUE THREAD

**Issue by andygrove** [apache/arrow-datafusion#3419]
**Describe the bug**
Arrow's `can_cast_types` has different logic than DataFusion's `coerce_types` and this results in bugs in type coercion.

**To Reproduce**

**Expected behavior**

**Additional context**

**Comment by liukun4515** [comment]
Do we need make align the type coercion with arrow?
The cast system should not be same with arrow, because the datafusion is the SQL system and has some special semantics than the arrow lib
@andygrove

**Comment by andygrove** [comment]
I agree that there will be differences. See the conversation at https://github.com/apache/arrow-datafusion/pull/3407#discussion_r967137788 for one example where we have to work around a difference and this example is possibly due to a bug in DataFusions logic

**Comment by andygrove** [comment]
I updated the title and description now that I understand this more.

**Comment by alamb** [comment]
Maybe we could return a struct like
```rust
enum NeededCoercion {
  coerce_inputs: Vec<Datatype>,
  output_type: DataType
}
```

Or something 🤔

**Comment by HaoYang670** [comment]
Hi, could someone help to explain what the `coerce_types` does. Does it coerce the lhs and rhs types, or return the `return type` of the operator?

**Comment by alamb** [comment]
> Hi, could someone help to explain what the coerce_types does. Does it coerce the lhs and rhs types, or calculate the return type of the operator?

I think this is a reasonable description of coercion:

https://github.com/apache/arrow-datafusion/blob/9c24a79118c0ae0aee480bf039385a06d240b499/datafusion/expr/src/type_coercion.rs#L18-L32
